File: tokenmon/sprites/convert.py
#!/usr/bin/env -S uv run --quiet python3
# /// script
# requires-python = ">=3.9"
# dependencies = ["Pillow"]
# ///
"""Convert a Pokémon PNG sprite to ▀▄ half-block ANSI 256-color terminal art."""

import argparse
import sys
import os

def ansi_256(r, g, b):
    """Convert RGB to nearest ANSI 256 color code."""
    # Grayscale range
    if r == g == b:
        if r < 8:
            return 16
        if r > 248:
            return 231
        return round((r - 8) / 247 * 24) + 232
    # 6x6x6 color cube
    ri = round(r / 255 * 5)
    gi = round(g / 255 * 5)
    bi = round(b / 255 * 5)
    return 16 + 36 * ri + 6 * gi + bi

def fg(r, g, b):
    return f"\x1b[38;5;{ansi_256(r,g,b)}m"

def bg(r, g, b):
    return f"\x1b[48;5;{ansi_256(r,g,b)}m"

RESET = "\x1b[0m"

def convert_image(img, width):
    """Convert PIL image to half-block ANSI art string."""
    # Resize maintaining aspect ratio (half-block: each char = 2 pixel rows)
    orig_w, orig_h = img.size
    height = int(orig_h * width / orig_w)
    if height % 2 != 0:
        height += 1
    img = img.resize((width, height), resample=1)  # LANCZOS=1

    lines = []
    for y in range(0, height, 2):
        line = ""
        for x in range(width):
            top_px = img.getpixel((x, y))
            bot_y = y + 1
            bot_px = img.getpixel((x, bot_y)) if bot_y < height else (0, 0, 0, 0)

            # Handle RGBA transparency
            if img.mode == "RGBA":
                tr, tg, tb, ta = top_px
                br, bg_, bb, ba = bot_px
                top_transparent = ta < 128
                bot_transparent = ba < 128
            else:
                tr, tg, tb = top_px[:3]
                br, bg_, bb = bot_px[:3]
                top_transparent = False
                bot_transparent = False

            if top_transparent and bot_transparent:
                line += " "
            elif top_transparent:
                line += f"{fg(br, bg_, bb)}▄{RESET}"
            elif bot_transparent:
                line += f"{fg(tr, tg, tb)}▀{RESET}"
            else:
                line += f"{fg(tr, tg, tb)}{bg(br, bg_, bb)}▀{RESET}"
        lines.append(line)
    return "\n".join(lines)

def text_fallback(name, width):
    """Generate a simple text placeholder."""
    label = f"[{name}]"
    return label.center(width)

def main():
    parser = argparse.ArgumentParser(description="Convert Pokémon sprite PNG to terminal ANSI art")
    parser.add_argument("--id", type=int, required=True, help="Pokédex ID")
    parser.add_argument("--width", type=int, default=20, help="Output width in characters")
    parser.add_argument("--input", type=str, help="Input PNG file path (overrides --id lookup)")
    parser.add_argument("--output", type=str, help="Output text file path (default: stdout)")
    parser.add_argument("--name", type=str, default="", help="Pokémon name for text fallback")
    args = parser.parse_args()

    # Determine input file
    if args.input:
        input_path = args.input
    else:
        # Look for sprite in standard locations relative to this script
        script_dir = os.path.dirname(os.path.abspath(__file__))
        input_path = os.path.join(script_dir, "raw", f"{args.id}.png")

    # Try Pillow conversion
    result = None
    try:
        from PIL import Image
        if not os.path.exists(input_path):
            raise FileNotFoundError(f"Sprite not found: {input_path}")
        img = Image.open(input_path)
        result = convert_image(img, args.width)
    except ImportError:
        # Pillow not available
        name = args.name or str(args.id)
        result = text_fallback(name, args.width)
    except FileNotFoundError:
        name = args.name or str(args.id)
        result = text_fallback(name, args.width)
    except Exception as e:
        name = args.name or str(args.id)
        result = text_fallback(name, args.width)

    # Output
    if args.output:
        os.makedirs(os.path.dirname(os.path.abspath(args.output)), exist_ok=True)
        with open(args.output, "w", encoding="utf-8") as f:
            f.write(result + "\n")
    else:
        print(result)

if __name__ == "__main__":
    main()

File: tokenmon/src/sprites/convert.ts
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'fs';
import { dirname } from 'path';
import { PNG } from 'pngjs';

/**
 * Convert RGB to nearest ANSI 256 color code.
 */
function ansi256(r: number, g: number, b: number): number {
  // Grayscale range
  if (r === g && g === b) {
    if (r < 8) return 16;
    if (r > 248) return 231;
    return Math.round((r - 8) / 247 * 24) + 232;
  }
  // 6x6x6 color cube
  const ri = Math.round(r / 255 * 5);
  const gi = Math.round(g / 255 * 5);
  const bi = Math.round(b / 255 * 5);
  return 16 + 36 * ri + 6 * gi + bi;
}

function fg(r: number, g: number, b: number): string {
  return `\x1b[38;5;${ansi256(r, g, b)}m`;
}

function bg(r: number, g: number, b: number): string {
  return `\x1b[48;5;${ansi256(r, g, b)}m`;
}

const RESET = '\x1b[0m';

interface RGBA {
  r: number;
  g: number;
  b: number;
  a: number;
}

/**
 * Convert a PNG image to half-block ANSI terminal art.
 * Uses pngjs for PNG parsing.
 */
export function convertPng(pngBuffer: Buffer, width: number = 20): string {
  const img = PNG.sync.read(pngBuffer);
  const origW = img.width;
  const origH = img.height;

  // Resize maintaining aspect ratio
  let targetH = Math.round(origH * width / origW);
  if (targetH % 2 !== 0) targetH += 1;

  // Simple nearest-neighbor resize
  const getPixel = (x: number, y: number): RGBA => {
    const srcX = Math.min(Math.floor(x * origW / width), origW - 1);
    const srcY = Math.min(Math.floor(y * origH / targetH), origH - 1);
    const idx = (srcY * origW + srcX) * 4;
    return {
      r: img.data[idx],
      g: img.data[idx + 1],
      b: img.data[idx + 2],
      a: img.data[idx + 3],
    };
  };

  const lines: string[] = [];
  for (let y = 0; y < targetH; y += 2) {
    let line = '';
    for (let x = 0; x < width; x++) {
      const top = getPixel(x, y);
      const bot = y + 1 < targetH ? getPixel(x, y + 1) : { r: 0, g: 0, b: 0, a: 0 };

      const topTransparent = top.a < 128;
      const botTransparent = bot.a < 128;

      if (topTransparent && botTransparent) {
        line += ' ';
      } else if (topTransparent) {
        line += `${fg(bot.r, bot.g, bot.b)}▄${RESET}`;
      } else if (botTransparent) {
        line += `${fg(top.r, top.g, top.b)}▀${RESET}`;
      } else {
        line += `${fg(top.r, top.g, top.b)}${bg(bot.r, bot.g, bot.b)}▀${RESET}`;
      }
    }
    lines.push(line);
  }

  return lines.join('\n');
}

export function textFallback(name: string, width: number): string {
  const label = `[${name}]`;
  const pad = Math.max(0, Math.floor((width - label.length) / 2));
  return ' '.repeat(pad) + label;
}

/**
 * Convert a PNG file to terminal art and optionally write to file.
 */
export function convertFile(inputPath: string, outputPath?: string, width: number = 20, name?: string): string {
  if (!existsSync(inputPath)) {
    return textFallback(name ?? '?', width);
  }

  try {
    const buf = readFileSync(inputPath);
    const result = convertPng(buf, width);

    if (outputPath) {
      mkdirSync(dirname(outputPath), { recursive: true });
      writeFileSync(outputPath, result + '\n', 'utf-8');
    }

    return result;
  } catch {
    return textFallback(name ?? '?', width);
  }
}
